### DISCUSSION TITLE
Using a custom installation path for node

### DISCUSSION THREAD

**Discussion by JosXa** [coreybutler/nvm-windows#1128] [General] [answered]
I'm working with a project that comes with a very specific installation and set of requirements for Node, and I'd like to add their NodeJS installation to nvm in order to be able to switch seamlessly. I have tried naively copying/symlinking the directory to `C:\ProgramData\nvm` in a new folder, but nvm-windows appears to keep its own registry of available Node versions.

Is there any way to link it to a custom installation?

**Comment by coreybutler** [comment, marked as answer]
There's no special registry. NVM4W reads the `NVM_HOME` directory, identifying node versions named with the `vX.X.X` pattern. If you want to install a specific version w/o using `nvm install`, just drop the appropriately named folder into `NVM_HOME`. As for creating the symlink, I'd recommend using `nvm use ...`, but if you really want to do it on your own, use mklink to update the symlink target of `NVM_SYMLINK`. This is how NVM4W does it behind the scenes, with the major difference being NVM4W automatically attempts to elevate the command, leverage UAC prompts when necessary, etc.

**Reply by JosXa** [reply]
Ah great, thank you so much!

**Reply by JosXa** [reply]
Seems to work quite well with a hardlink (junction).
One interesting caveat: `nvm use <hardlinked>` works, but `nvm list` and `nvm current` then do not make this custom version appear in the output:

```
$ nvm use 20.11.1
Now using node v20.11.1 (64-bit)
$ node --version
v20.11.1
$ nvm current
No current version. Run 'nvm use x.x.x' to set a version.
$ nvm list

    21.7.3
    20.12.2
    20.4.0
    18.18.0
    18.17.1
    16.14.0
```

And one follow-up question: Is there any room in the version parsing regex that would allow me to somehow mark this node installation as custom? I tried naming the folder `v20.11.1-knode` and `v20.11.1.knode`, but those don't parse. I guess I'm asking for a hack ;)

**Reply by coreybutler** [reply]
There's nothing in the parser that would indicate a custom installation. 

Generally speaking, NVM4W doesn't have a design goal to support that type of workflow. Custom installs are something we're looking into for [Runtime](https://github.com/coreybutler/nvm-windows/wiki/Runtime). I'd be open to hearing more about the use case and the general expectations/desires around this (i.e. what would you do with a custom version if you could identify it?)

**Reply by JosXa** [reply]
TIL about Runtime, sounds like a very useful addition to the ecosystem 🥳 

The only use case I've come across and the one this discussion is about is [Script Kit](https://scriptkit.com)'s custom "knode" installation. The installer puts a node executable in your home folder, `~/.knode/bin`, and requires you to use exactly that version when interacting with Script Kit.
The need to identify it would be twofold:  
- Being able to install the official version of the node runtime simultaneously (rare)
- Being able to switch to that knode environment with a nice alias, as it's very common that you have to switch back and forth.

All in all not very urgent requests, as these problems can be solved with custom scripts to switch to that specific version (e.g. I'm using Kit itself for that).

**Comment by JosXa** [comment]
Just had the epiphany that I can just name the folder v1.1.1 or some other magic number to have a good alias

**Comment by kenwarner** [comment]
didn't see whether this had been debated in other discussions/issues/PRs but I noticed in the 1.2.0 release notes

> The official installer now defaults to C:\nvm4w\nodejs for the NVM_SYMLINK instead of C:\Program Files\nodejs. Windows has the concept of “protected directories” which require elevated permissions to access. C:\Program Files is one of these directories. Users with limited permissions post-installation are less likely to encounter access issues with this alternative directory. There is no impact on privileged user accounts.

I get the explanation re: Program Files, but I don't think many will be a big fan of the choice to create a folder at the root of the C:\ drive. Any reason not to use AppData or some other location by default?

**Reply by coreybutler** [reply]
@kenwarner the main concern with AppData is that it is user specific, unless you install to the public user AppData. In shared environments, this can be a challenge. I've reached the conclusion that not everyone will be happy, but the installer gives you the option to customize. That's the best balance I can think of.
